x/qgb/abci.go

```diff
@@ -113,3 +122,118 @@ func handleValsetRequest(ctx sdk.Context, k keeper.Keeper) {
+		return
+	}
+	if lastNonceHeight > lastUnbondingHeight {
+		// checking if it's the initial case following the startup of the chain
+		if lastNonceHeight == 1 && lastUnbondingHeight == 0 {
+			return
+		}
+		// we should never hit this case, since we will keep the attestations up to the last unbonding height
+		panic("missing attestations up to the unbonding height")
+	}
+	// now we have attestations before the unbonding height, we should check whether we need to prune them
+	// or not yet
+	latestAttestationNonce := k.GetLatestAttestationNonce(ctx)
+	if latestAttestationNonce-lastAvailableNonce+1 <= AttestationPruningThreshold {
+		// we don't need to prune as we still have room to store more attestations
+		return
+	}
+	// now we want to prune attestations as we have the following conditions:
+	// - we have attestations up to the last unbonding height
+	// - the total number of attestations, including the ones before the unbonding height, are greater
+	// than the AttestationPruningThreshold
+	err := pruneAttestations(ctx, k, lastAvailableNonce, latestAttestationNonce)
+	if err != nil {
+		panic(err)
+	}
+}
+
+// getLastAvailableNonceHeight returns the block height corresponding to the last available
+// nonce in store.
+func getLastAvailableNonceHeight(ctx sdk.Context, k keeper.Keeper) uint64 {
+	lastAttestationInStore, found, err := k.GetAttestationByNonce(ctx, k.GetLastAvailableAttestationNonce(ctx))
+	if err != nil {
+		panic(err)
+	}
+	if !found {
+		panic(fmt.Sprintf("couldn't find attestation %s for pruning", lastAttestationInStore))
+	}
+	switch lastAttestationInStore.Type() {
+	case types.DataCommitmentRequestType:
+		dc := lastAttestationInStore.(*types.DataCommitment)
+		return dc.EndBlock
+	case types.ValsetRequestType:
+		vs := lastAttestationInStore.(*types.Valset)
+		return vs.Height
+	default:
+		panic("unknown attestation type")
+	}
+}
+
+// pruneAttestations prunes attestations to keep attestations up to the last unbonding height
+// and the total number of attestations lower than the AttestationPruningThreshold.
+func pruneAttestations(
+	ctx sdk.Context,
+	k keeper.Keeper,
+	lastAvailableAttestationNonce uint64,
+	latestAttestationNonce uint64,
+) error {
+	ctx.Logger().Debug("pruning attestations from qgb store")
+	if !k.CheckLastUnbondingNonce(ctx) {
+		// We should never hit this case in the happy path because at this level, we're sure that
+		// all store values are initialized.
+		// However, we might hit it in an upgrade if we don't handle the store values carefully.
+		// So, it's good to have this check in place.
+		return fmt.Errorf("last unbonding nonce not initialized in store")
+	}
+	lastUnbondingNonce := k.GetLastUnbondingNonce(ctx)
+	count := 0
+	newLastAvailableNonce := lastAvailableAttestationNonce
+	for i := lastAvailableAttestationNonce; i < lastUnbondingNonce; i++ {
+		if newLastAvailableNonce == lastUnbondingNonce ||
```

Comment:
I don't see how this clause ever hits. `lastUnbondingNonce` is always less than `i` and thus `newLastAvailableNonce`

Reply:
21e373f, also found more simplifications when updating